lib/teiserver/lobby/libs/lobby_restrictions.ex

```diff
@@ -0,0 +1,293 @@
+
+    # Chevlevel stuff here
+    cond do
+      # Default chev levels
+      max_rank_to_play >= @rank_upper_bound &&
+          min_rank_to_play <= 0 ->
+        nil
+
+      # Just a max rating
+      max_rank_to_play < @rank_upper_bound &&
+          min_rank_to_play <= 0 ->
+        "Max chev: #{max_rank_to_play + 1}"
+
+      # Just a min rating
+      max_rank_to_play >= @rank_upper_bound &&
+          min_rank_to_play > 0 ->
+        "Min chev: #{min_rank_to_play + 1}"
+
+      # Chev range
+      # It shouldn't go here
+      max_rank_to_play < @rank_upper_bound ||
+          min_rank_to_play > 0 ->
+        "Chev between: #{min_rank_to_play} - #{max_rank_to_play}"
+
+      true ->
+        nil
+    end
+  end
+
+  def get_rating_bounds_for_title(consul_state) when consul_state == nil do
+    nil
+  end
+
+  def get_rating_bounds_for_title(consul_state) do
+    max_rating_to_play = Map.get(consul_state, :maximum_rating_to_play, @rating_upper_bound)
+    min_rating_to_play = Map.get(consul_state, :minimum_rating_to_play, 0)
+
+    cond do
+      # Default ratings
+      max_rating_to_play >= @rating_upper_bound &&
+          min_rating_to_play <= 0 ->
+        nil
+
+      # Just a max rating
+      max_rating_to_play < @rating_upper_bound &&
+          min_rating_to_play <= 0 ->
+        "Max rating: #{max_rating_to_play}"
+
+      # Just a min rating
+      max_rating_to_play >= @rating_upper_bound &&
+          min_rating_to_play > 0 ->
+        "Min rating: #{min_rating_to_play}"
+
+      # Rating range
+      max_rating_to_play < @rating_upper_bound ||
+          min_rating_to_play > 0 ->
+        "Rating between: #{min_rating_to_play} - #{max_rating_to_play}"
+
+      true ->
+        nil
+    end
+  end
+
+  def get_failed_rank_check_text(player_rank, consul_state) do
+    bounds = get_rank_bounds_for_title(consul_state)
+
+    [
+      @splitter,
+      "You don't meet the chevron requirements for this lobby (#{bounds}). Your chevron level is #{player_rank + 1}. Learn more about chevrons here:",
+      "https://www.beyondallreason.info/guide/rating-and-lobby-balance#rank-icons"
+    ]
+  end
+
+  def get_failed_rating_check_text(player_rating, consul_state, rating_type) do
+    bounds = get_rating_bounds_for_title(consul_state)
+    player_rating_text = player_rating |> Decimal.from_float() |> Decimal.round(2)
+
+    [
+      @splitter,
+      "You don't meet the rating requirements for this lobby (#{bounds}). Your #{rating_type} match rating is #{player_rating_text}. Learn more about rating here:",
+      "https://www.beyondallreason.info/guide/rating-and-lobby-balance#openskill"
+    ]
+  end
+
+  @spec check_rank_to_play(any(), any()) :: :ok | {:error, iodata()}
+  def check_rank_to_play(user, consul_state) do
+    state = consul_state
+
+    # Contributors auto pass since their ranks are not defined on playtime. To be fixed seperately.
+    is_contributor? = CacheUser.is_contributor?(user)
```

Comment:
Rank is calculated in Teiserver.CacheUser, however there are alternative ways to calculate rank which can be set with `profile.Rank method` site configuration.

The default one and the one used on server4 is `Role`
```ex
  def calculate_rank(userid, "Role") do
    ingame_hours = rank_time(userid)

    cond do
      has_any_role?(userid, ~w(Core Contributor)) -> 6
      ingame_hours > 1000 -> 5
      ingame_hours > 250 -> 4
      ingame_hours > 100 -> 3
      ingame_hours > 15 -> 2
      ingame_hours > 5 -> 1
      true -> 0
    end
  end
```

Other options are `Uncertainity`, `Leaderboard rating` and `Playtime`
```ex
  def calculate_rank(userid, "Playtime") do
    ingame_hours = rank_time(userid)

    [5, 15, 30, 100, 300, 1000, 3000]
    |> Enum.count(fn r -> r <= ingame_hours end)
  end

  # Using leaderboard rating
  def calculate_rank(userid, "Leaderboard rating") do
    rating = Account.get_player_highest_leaderboard_rating(userid)

    [3, 7, 12, 21, 26, 35, 1000]
    |> Enum.count(fn r -> r <= rating end)
  end

  def calculate_rank(userid, "Uncertainty") do
    uncertainty =
      Account.get_player_lowest_uncertainty(userid)
      |> :math.ceil()

    (8 - uncertainty)
    |> max(0)
    |> max(7)
  end
  ```

`Role` is the only one where contributors get a special rank so it might be a good idea to add a check for current rank method here as well.

Reply:
Updated